File: typescript/src/booking/integration/giftcard/bookingResultIntegrationEvent.ts
// Published Language esposta dal BC booking verso il BC giftcard.
// Rappresenta il risultato di una prenotazione senza esporre gli eventi di dominio interni di booking.

import type { Money } from '@/common/domain/primitive/money.js';

export type BookingCompletedIntegrationEvent = {
  readonly kind: 'BookingCompletedIntegrationEvent';
  readonly giftCardReference: string;
  readonly amount: Money;
};

export type BookingRefusedIntegrationEvent = {
  readonly kind: 'BookingRefusedIntegrationEvent';
  readonly giftCardReference: string;
  readonly amount: Money;
};

export type BookingRejectedIntegrationEvent = {
  readonly kind: 'BookingRejectedIntegrationEvent';
  readonly giftCardReference: string;
  readonly amount: Money;
};

export type BookingResultIntegrationEvent =
  | BookingCompletedIntegrationEvent
  | BookingRefusedIntegrationEvent
  | BookingRejectedIntegrationEvent;

export function bookingCompletedIntegrationEvent(
  giftCardReference: string,
  amount: Money
): BookingCompletedIntegrationEvent {
  return { kind: 'BookingCompletedIntegrationEvent', giftCardReference, amount };
}

export function bookingRefusedIntegrationEvent(
  giftCardReference: string,
  amount: Money
): BookingRefusedIntegrationEvent {
  return { kind: 'BookingRefusedIntegrationEvent', giftCardReference, amount };
}

export function bookingRejectedIntegrationEvent(
  giftCardReference: string,
  amount: Money
): BookingRejectedIntegrationEvent {
  return { kind: 'BookingRejectedIntegrationEvent', giftCardReference, amount };
}

File: java/src/main/java/it/giannibombelli/wsc2026/booking/integration/giftcard/BookingResultIntegrationEvent.java
package it.giannibombelli.wsc2026.booking.integration.giftcard;

import it.giannibombelli.wsc2026.common.domain.primitive.Money;

import java.util.UUID;

/**
 * Published Language esposta dal BC {@code booking} verso il BC {@code giftcard}.
 * <p>
 * Rappresenta il risultato di una prenotazione senza esporre gli eventi di dominio interni di {@code booking}.
 * I campi sono espressi con tipi stabili e condivisi (UUID, Money) per mantenere il contratto leggero e indipendente
 * dall'implementazione dei due BC.
 */
public sealed interface BookingResultIntegrationEvent {

    record BookingCompletedIntegrationEvent(UUID giftCardReference, Money amount) implements BookingResultIntegrationEvent {
    }

    record BookingRefusedIntegrationEvent(UUID giftCardReference, Money amount) implements BookingResultIntegrationEvent {
    }

    record BookingRejectedIntegrationEvent(UUID giftCardReference, Money amount) implements BookingResultIntegrationEvent {
    }
}
